### ISSUE TITLE
Auto-discovery of flannel CNI and ClusterCIDR is failing on K3s

### ISSUE THREAD

**Issue by ccwalterhk** [submariner-io/submariner#1642]
**What happened**: Auto-discovery of CNI is failing on K3s. It does not detect the CNI in the cluster and uses "generic" route-agent.

```
 ✓ Showing Gateways
NODE                            HA STATUS       SUMMARY                         
k3-subctl-w1                    active          All connections (1) are established

    Discovered network details via Submariner:
        Network plugin:  generic
        Service CIDRs:   [10.45.0.0/16]
        Cluster CIDRs:   [10.44.0.0/24]
 ✓ Showing Network details
```

**What you expected to happen**: It should be able to detect flannel CNI.

**How to reproduce it (as minimally and precisely as possible)**:
Follow the exact steps in URL https://submariner.io/getting-started/quickstart/k3s/.  below is exact steps to build two clusters

```
Cluster a:
POD_CIDR=10.44.0.0/16
SERVICE_CIDR=10.45.0.0/16
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--cluster-cidr $POD_CIDR --service-cidr $SERVICE_CIDR" sh -s -

sudo snap install yq

sudo su

cp /etc/rancher/k3s/k3s.yaml kubeconfig.cluster-a
export IP=$(hostname -I | awk '{print $1}')
yq -i eval \
'.clusters[].cluster.server |= sub("127.0.0.1", env(IP)) | .contexts[].name = "cluster-a" | .current-context = "cluster-a"' \
kubeconfig.cluster-a

Cluster b:

POD_CIDR=10.144.0.0/16
SERVICE_CIDR=10.145.0.0/16
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--cluster-cidr $POD_CIDR --service-cidr $SERVICE_CIDR" sh -s -

sudo snap install yq

sudo su

cp /etc/rancher/k3s/k3s.yaml kubeconfig.cluster-b
export IP=$(hostname -I | awk '{print $1}')
yq -i eval \
'.clusters[].cluster.server |= sub("127.0.0.1", env(IP)) | .contexts[].name = "cluster-b" | .current-context = "cluster-b"' \
kubeconfig.cluster-b
```

**Anything else we need to know?**:

**Environment**:
- Diagnose information (use `subctl diagnose all`):

```
walter@k3-subctl-m1:~$ subctl show all --kubeconfig kubeconfig.cluster-a
Cluster "default"
 ✓ Showing Connections
GATEWAY       CLUSTER    REMOTE IP     NAT  CABLE DRIVER  SUBNETS                       STATUS     RTT avg.  
k3-subctl-w2  cluster-b  192.168.1.93  no   libreswan     10.145.0.0/16, 10.144.0.0/24  connected            

 ✓ Showing Endpoints
CLUSTER ID                    ENDPOINT IP     PUBLIC IP       CABLE DRIVER        TYPE            
cluster-a                     192.168.1.92    58.182.182.152  libreswan           local           
cluster-b                     192.168.1.93    58.182.182.152  libreswan           remote          

 ✓ Showing Gateways
NODE                            HA STATUS       SUMMARY                         
k3-subctl-w1                    active          All connections (1) are established

    Discovered network details via Submariner:
        Network plugin:  generic
        Service CIDRs:   [10.45.0.0/16]
        Cluster CIDRs:   [10.44.0.0/24]
 ✓ Showing Network details

COMPONENT                       REPOSITORY                                            VERSION         
submariner                      quay.io/submariner                                    0.11.0          
submariner-operator             quay.io/submariner                                    0.11.0          
service-discovery               quay.io/submariner                                    0.11.0          
 ✓ Showing versions

walter@k3-subctl-m1:~$ subctl show all --kubeconfig kubeconfig.cluster-b
Cluster "default"
 ✓ Showing Connections
GATEWAY       CLUSTER    REMOTE IP     NAT  CABLE DRIVER  SUBNETS                     STATUS     RTT avg.  
k3-subctl-w1  cluster-a  192.168.1.92  no   libreswan     10.45.0.0/16, 10.44.0.0/24  connected            

 ✓ Showing Endpoints
CLUSTER ID                    ENDPOINT IP     PUBLIC IP       CABLE DRIVER        TYPE            
cluster-b                     192.168.1.93    58.182.182.152  libreswan           local           
cluster-a                     192.168.1.92    58.182.182.152  libreswan           remote          

 ✓ Showing Gateways
NODE                            HA STATUS       SUMMARY                         
k3-subctl-w2                    active          All connections (1) are established

    Discovered network details via Submariner:
        Network plugin:  generic
        Service CIDRs:   [10.145.0.0/16]
        Cluster CIDRs:   [10.144.0.0/24]
 ✓ Showing Network details

COMPONENT                       REPOSITORY                                            VERSION         
submariner                      quay.io/submariner                                    0.11.0          
submariner-operator             quay.io/submariner                                    0.11.0          
service-discovery               quay.io/submariner                                    0.11.0          
 ✓ Showing versions

walter@k3-subctl-m1:~$ 
```

- Gather information (use `subctl gather`):
[submariner-cluster-A-20211227124333.tar.gz](https://github.com/submariner-io/submariner/files/7784854/submariner-cluster-A-20211227124333.tar.gz)
[submariner-cluster-B-20211227124346.tar.gz](https://github.com/submariner-io/submariner/files/7784852/submariner-cluster-B-20211227124346.tar.gz)

- Cloud provider or hardware configuration:
VMWARE fusion to build K3S

- Install tools:
NA
- Others:
NA

**Comment by nyechiel** [comment]
Thanks for reporting this @cwalterhk. IIRC, despite the fact that the CNI is not being discovered properly, actual routing logic is in place and connectivity works as expected. Can you confirm? /cc @sridhargaddam

**Comment by ccwalterhk** [comment]
> Thanks for reporting this @cwalterhk. IIRC, despite the fact that the CNI is not being discovered properly, actual routing logic is in place and connectivity works as expected. Can you confirm? /cc @sridhargaddam

Confirm that the routing logic is place. After I forced the clusterCIDR, everything works.

**Comment by sridhargaddam** [comment]
@nyechiel yes Nir, the routing logic would work fine. However, apart from CNI detection failure, the other problem is that Submariner operator was not detecting the clusterCIDR properly and wrong CIDR was getting configured in the cluster which was leading to the datapath failures.

**Comment by yboaron** [comment]
IIUC, there are two separate issues here:
1. add support for flannel plugin discovery (currently flannel discovered as generic)
2. wrong pod cidr is retrieved when Submariner discovered flannel as generic plugin.

'2' is tracked at https://github.com/submariner-io/submariner-operator/issues/1730.
As per '1', I think we should handle it as a feature/enhancement request and not a bug.

**Comment by nerdalert** [comment]
Took a stab at Flannel discovery with in [#1850](https://github.com/submariner-io/submariner-operator/pull/1850) in operator

**Comment by nyechiel** [comment]
Thanks @nerdalert! As we are entering the final stages of the 0.12 cycle and this issue is not urgent AFAICT, is it ok if we defer this to 0.13? Just trying to make sure we meet our 0.12 priorities first. /cc @sridhargaddam @skitt FYI

**Comment by nerdalert** [comment]
Sounds good @nyechiel! Generic discovery covers the functionality for now. I mentioned in the PR for flannel discovery it is post v0.12. Maybe I should have put v0.13 instead.

**Comment by josepowera** [comment]
At least update https://submariner.io/getting-started/quickstart/k3s/ with requirement for additional parameter until this is fixed 
--clustercidr "10.44.0.0/16" or similar.   Got here the hard way after long time debugging.

**Comment by dfarrell07** [comment]
> At least update https://submariner.io/getting-started/quickstart/k3s/ with requirement for additional parameter until this is fixed --clustercidr "10.44.0.0/16" or similar. Got here the hard way after long time debugging.

Thanks, we'll look at updating the docs :+1:

**Comment by tpantelis** [comment]
Looks like this is fixed by https://github.com/submariner-io/submariner-operator/pull/1850 which will be in 0.14.
